### DISCUSSION TITLE
Cycle 5: Search (15 Feb 2021 - 31 Mar 2021)

### DISCUSSION THREAD

**Discussion by levogirar** [wuwanahq/wuwana2#122] [Cycles]
# Cycle 5: Search (15 Feb 2021 - 31 Mar 2021)

## Problem
In this cycle, we will focus on implementing the search function in the website. This entails a version jump to v2.2 [Github Project](https://github.com/wuwanahq/wuwana2/projects/1).

We stopped using the basic category filter since it was very rigid and doesn't work well with fuzzy edges (like how to filter specialty coffee roasters from regular coffee roasters?). 

Then we decided instead of manually labelling each individual company by implementing a system of tagging and keyword matching. The result was not satisfactory as described in #94. Also, this did not solve the retrieving issue. 

After implementing the NLP to tag companies, we have higher quality tags compared to our previous attempts. Now we need a way to retrieve the companies.

## Solution
We decided to implement a basic search box, similar to any other search engine. We will not use any third-party service and recreate this by ourselves as much as possible. 

### Implementation

* ￼The new homepage will be a search box with minor instructions and an about Wuwana section.
* Results will be selected based on **text-match** from company name and tags. 
* The current homepage will become the search result page.

**Flow**

1. Users land on the homepage.
2. Users make a search query.
3. Users can only edit the search from the search page (users only go to the homepage when: clicking on the home button, go back from the browser, or change the URL to wuwana.com).

### Limitations and out of scope

1. **Pre-generated search suggestions are out of scope** (ie. Showing coffee roasters, breweries, company name, etc under the search box). We have not defined the criteria for them. Maybe in the future. However, we want to limit the "blank page" effect, so we will have a write-up giving them suggestions of what they can search for.
2. **Sort results by internal relevance metrics**. By default, we will show the most recently updated (or created) company first.

### Design
The design v2.2 can be found in Figma. @levogirar will design it and implement it. On Figma, the design is for reference, the actual output might be different but the general shape is there.

* https://www.figma.com/file/ypwyPKukse5zbQUMjbGakS/wuwana-v2?node-id=934%3A5354

## Other changes and work

**Mobile sidebar:**
We will implement a better sidebar following Google's suggestion [https://web.dev/building-a-sidenav-component/]. 

**Mobile company profile page:**
We will make the company main information mobile-friendly.

**Moving the language selection to the footer**

**Admin login session:** This is needed for admin to edit the company information.

**Website analytics:** This includes overall website traffic, company page visit, referrals, etc.

## Appetite
6 weeks

## Things to consider
- How are we going to handle misspelling and typos?
- What about searches in different languages?
- What is the backend search flow diagram?
- What happen with no-result serches?

**Comment by Nils85** [comment]
We definitely need to shift the planning for this cycle! 😳
Because we are gonna start the version 2.2 around March 15th so...

**Reply by levogirar** [reply]
But we have until March 31st to do the search feature, no? The design is mostly done, we need the back-end to support it.

**Reply by Nils85** [reply]
It depends when we are gonna start this version 2.2. We will see...

**Comment by Nils85** [comment]
Hey @levogirar what the folder `wuwana2/admin` is for?

**Reply by levogirar** [reply]
@Nils85 I thing it is the left over from the merging the v2.1 to v2.2. I deleted them from my computers and pushed into v22-design. I think I didnt it is not reflected on the develop-v22, tho. You could ignore it and/or delete it.

**Comment by Nils85** [comment]
There is two different versions for the same HTML code representing a company cards.

One in `homepage/view`:

```
<a class="card" href="/...">
	<div class="logo-main margin-r16">
		<img src="..." alt="logo" onerror="setDefaultImg()">
	</div>
	<div class="company-card-wrapper">
		<div class="company-card-info">
			<h3>...</h3>
			<ul><li>...</li></ul>
			<div class="button-icon-small margin-t-auto">
				<img src="/static/icon/tiny/map.svg" alt="">...
			</div>
		</div>
	</div>
</a>
```

And another in `ajax/pagination`:

```
<a class="card" href="/...">
	<div class="logo-main margin-r16">
		<img src="..." alt="logo" onerror="setDefaultImg()">
	</div>
	<div class="company-card-wrapper">
		<div class="company-card-info">
			<h3>...</h3>
			<ul class="tag-area">
				<li>...</li>
			</ul>
			<div class="button-icon-small margin-t-auto">
				<img src="/static/icon/tiny/map.svg" alt="">...
			</div>
		</div>
		<div class="company-card-badge-wrapper"></div>
	</div>
</a>
```

Which one should we use in all situations?
Is `<div class="company-card-badge-wrapper"></div>` still used?
And what about the CSS class `tag-area` for the UL element?

**Reply by levogirar** [reply]
We should use the homepage/view one. Both `tag-area` and `company-card-badge-wrapper` have been depreciated.

**Comment by Nils85** [comment]
This is the **current** backend search flow and how the result will be displayed:

- The user type a search then sends its request
- The backend receives the search then requests the database
- Get result set 1 that match company names then order it by the most recent updated
- Get result set 2 that match first tag names then order it by the most recent updated
- Get result set 3 that match second tag names then order it by the most recent updated
- Display result set in this order 1, 2 then 3

Note: searching on tag names works with the 4 languages available. 
You can see this flow in action by searching "eco" for example:

<img width="368" alt="Screen Shot 2021-03-27 at 1 10 59 PM" src="https://user-images.githubusercontent.com/11949055/112720335-f4364700-8efd-11eb-8d06-a1fef719261f.png">

- The first company comes from the result set 1 because it matched the name
- The second company comes from the result set 2 because it matched the first tag name
- The rest comes from the result set 3 because it matched the second tag name and is ordered by the most recent updated

It's possible to add more result sets to display more companies.
For example, result set 4 search on "other tags", result set 5 search on "description", etc...
So I think it should be a good idea to add a lot of result sets to be sure to have result in almost all cases. And the number of results is not going to a problem since we have a paging system on the result page.

**What do you think about this search flow?**
**Any suggestions to improve it?**

**Reply by Nils85** [reply]
By the way, I don't know why companies logo doesn't appear anymore... 🤔

**Reply by levogirar** [reply]
I like the process. Maybe in the future we could improve it with spelling mistakes and auto-suggestions. For now, I think it works very well. 

I also like the idea of going through the description and other tags. Although, I don't know how the nlptagger will affect it. 

For the logo, I think this happens because none of the companies have been updated recently, and the Instagram  link doesn't work. I am still able to see some logos on my side.

**Comment by levogirar** [comment]
@Nils85 The pagination doesn't work after a search. I've tried both on Chrome and Safari.

**Reply by Nils85** [reply]
It works on my side... Have you try to clear your browser cache? What did you searched?

**Reply by levogirar** [reply]
I cleaned everything from the browser, but I still can't click on view more. It doesn't work for any search 😢

**Reply by levogirar** [reply]
I re-tested it today. It works when I search for something but it doesn't when I simply click on the search button without searching for anything. 

![Screen Shot 2021-03-28 at 10 33 16](https://user-images.githubusercontent.com/54992530/112755995-2b196500-8fb1-11eb-8ed1-1ce7c0deecd6.png)
![Screen Shot 2021-03-28 at 10 33 41](https://user-images.githubusercontent.com/54992530/112756001-2ce32880-8fb1-11eb-9c78-8c2664849a36.png)
![Screen Shot 2021-03-28 at 10 34 16](https://user-images.githubusercontent.com/54992530/112756002-2e145580-8fb1-11eb-9414-0179f8a14b83.png)

**Reply by Nils85** [reply]
This is not supposed to happen, so I created the issue #167

**Comment by Nils85** [comment]
Hey @levogirar, have you decided about how to handle the problem with dead image links?
This is the solutions we found together:
- Hide outdated logo with some JavaScript (current solution)
- Save all logo and pictures on the server to always have somethings to display
- Maybe you had another idea since last time?

**Reply by levogirar** [reply]
I like the idea of saving the logo. It could help with the OG metatags and the the twiiter cards. 

Currently, the JS solution works partially. Sometimes the `onerror` eventListener doesn't work 🤷🏻‍♂️. 

I didn't think of a different solution for this problem besides the JS solution.

**Reply by Nils85** [reply]
So I created the issue #179

**Reply by levogirar** [reply]
Great!

**Comment by levogirar** [comment]
@Nils85 When do you think we could merge the develop branch to the main branch? On the project, it seems that we are only missing the analytics tools [optional] and the nlptagger. For the next cycle, I want to focus on adding data to the website. So we can work on fixing instagram, adding google places, improve the crawler, and work on the facebook review scraper.

**Reply by Nils85** [reply]
Did you try to add new companies with this branch?

**Reply by levogirar** [reply]
It seems that I am not able to create new companies. Perhaps a combination of the being blocked by Instagram and how we create a new company?
